File: app/src/main/java/medlife/com/br/helper/UsuarioFirebase.java
package medlife.com.br.helper;


import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;
import com.google.firebase.firestore.FirebaseFirestore;

public class UsuarioFirebase {

    public static String getIdUsuario(){

        FirebaseAuth autenticacao = ConfiguracaoFirebase.getFirebaseAutenticacao();
        return autenticacao.getCurrentUser().getUid();

    }

    public static FirebaseUser getUsuarioAtual(){
        FirebaseAuth usuario = ConfiguracaoFirebase.getFirebaseAutenticacao();
        return  usuario.getCurrentUser();
    }


    /**
     * Update the lastLogin timestamp for the current user
     */
    public static void atualizarLastLogin() {
        FirebaseUser currentUser = getUsuarioAtual();
        if (currentUser != null) {
            FirebaseFirestore db = FirebaseFirestore.getInstance();
            db.collection("usuarios").document(currentUser.getUid())
                    .update("lastLogin", com.google.firebase.Timestamp.now())
                    .addOnSuccessListener(aVoid -> {
                        System.out.println("LastLogin updated successfully for user: " + currentUser.getUid());
                    })
                    .addOnFailureListener(e -> {
                        // Log error but don't block the app
                        System.out.println("Error updating lastLogin: " + e.getMessage());
                        e.printStackTrace();
                    });
        } else {
            System.out.println("No authenticated user found for lastLogin update");
        }
    }
    
    /**
     * Update the lastLogin timestamp for a specific user
     */
    public static void atualizarLastLogin(String userId) {
        if (userId != null && !userId.isEmpty()) {
            FirebaseFirestore db = FirebaseFirestore.getInstance();
            db.collection("usuarios").document(userId)
                    .update("lastLogin", com.google.firebase.Timestamp.now())
                    .addOnSuccessListener(aVoid -> {
                        System.out.println("LastLogin updated successfully for user: " + userId);
                    })
                    .addOnFailureListener(e -> {
                        // Log error but don't block the app
                        System.out.println("Error updating lastLogin for user " + userId + ": " + e.getMessage());
                        e.printStackTrace();
                    });
        } else {
            System.out.println("Invalid userId for lastLogin update: " + userId);
        }
    }

    /**
     * Safely update lastLogin without overwriting other user data
     */
    public static void atualizarLastLoginSeguro(String userId) {
        if (userId != null && !userId.isEmpty()) {
            FirebaseFirestore db = FirebaseFirestore.getInstance();
            
            // First check if user document exists
            db.collection("usuarios").document(userId)
                    .get()
                    .addOnSuccessListener(documentSnapshot -> {
                        if (documentSnapshot.exists()) {
                            // User exists, safely update only lastLogin
                            db.collection("usuarios").document(userId)
                                    .update("lastLogin", com.google.firebase.Timestamp.now())
                                    .addOnSuccessListener(aVoid -> {
                                        System.out.println("LastLogin updated safely for user: " + userId);
                                    })
                                    .addOnFailureListener(e -> {
                                        System.out.println("Error updating lastLogin safely: " + e.getMessage());
                                    });
                        } else {
                            System.out.println("User document doesn't exist for lastLogin update: " + userId);
                        }
                    })
                    .addOnFailureListener(e -> {
                        System.out.println("Error checking user document for lastLogin update: " + e.getMessage());
                    });
        } else {
            System.out.println("Invalid userId for safe lastLogin update: " + userId);
        }
    }

}

File: app/src/main/java/medlife/com/br/helper/UsuarioFirebase.kt
package medlife.com.br.helper

import com.google.firebase.Timestamp
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.FirebaseFirestore

object UsuarioFirebase {
    @JvmStatic
    val idUsuario: String?
        get() {
            val autenticacao: FirebaseAuth = ConfiguracaoFirebase.firebaseAutenticacao
            return autenticacao.currentUser?.uid
        }

    @JvmStatic
    val usuarioAtual: FirebaseUser?
        get() {
            val usuario: FirebaseAuth = ConfiguracaoFirebase.firebaseAutenticacao
            return usuario.currentUser
        }

    /**
     * Update the lastLogin timestamp for the current user
     */
    @JvmStatic
    fun atualizarLastLogin() {
        val currentUser = usuarioAtual
        if (currentUser != null) {
            val db = FirebaseFirestore.getInstance()
            db.collection("usuarios").document(currentUser.uid)
                .update("lastLogin", Timestamp.now())
                .addOnSuccessListener {
                    println("LastLogin updated successfully for user: ${currentUser.uid}")
                }
                .addOnFailureListener { e ->
                    println("Error updating lastLogin: ${e.message}")
                    e.printStackTrace()
                }
        } else {
            println("No authenticated user found for lastLogin update")
        }
    }

    /**
     * Update the lastLogin timestamp for a specific user
     */
    @JvmStatic
    fun atualizarLastLogin(userId: String?) {
        if (!userId.isNullOrEmpty()) {
            val db = FirebaseFirestore.getInstance()
            db.collection("usuarios").document(userId)
                .update("lastLogin", Timestamp.now())
                .addOnSuccessListener {
                    println("LastLogin updated successfully for user: $userId")
                }
                .addOnFailureListener { e ->
                    println("Error updating lastLogin for user $userId: ${e.message}")
                    e.printStackTrace()
                }
        } else {
            println("Invalid userId for lastLogin update: $userId")
        }
    }

    /**
     * Safely update lastLogin without overwriting other user data
     */
    @JvmStatic
    fun atualizarLastLoginSeguro(userId: String?) {
        if (!userId.isNullOrEmpty()) {
            val db = FirebaseFirestore.getInstance()

            db.collection("usuarios").document(userId)
                .get()
                .addOnSuccessListener { documentSnapshot ->
                    if (documentSnapshot.exists()) {
                        db.collection("usuarios").document(userId)
                            .update("lastLogin", Timestamp.now())
                            .addOnSuccessListener {
                                println("LastLogin updated safely for user: $userId")
                            }
                            .addOnFailureListener { e ->
                                println("Error updating lastLogin safely: ${e.message}")
                            }
                    } else {
                        println("User document doesn't exist for lastLogin update: $userId")
                    }
                }
                .addOnFailureListener { e ->
                    println("Error checking user document for lastLogin update: ${e.message}")
                }
        } else {
            println("Invalid userId for safe lastLogin update: $userId")
        }
    }
}
